File: leet-code/349. Intersection of Two Arrays/IntersectionOfTwoArrays349_3.java
package com.example;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class IntersectionOfTwoArrays349_3 {
    public int[] intersection(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        int i = 0, j = 0;
        Set<Integer> set = new HashSet<>();
        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] == nums2[j]) {
                set.add(nums1[i]);
                i += 1;
                j += 1;
            } else if (nums1[i] < nums2[j]) {
                i += 1;
            } else {
                j += 1;
            }
        }
        return set.stream()
                .mapToInt(Integer::intValue)
                .toArray();
    }
}

File: leet-code/349. Intersection of Two Arrays/IntersectionOfTwoArrays349_3.kt
package com.example

class IntersectionOfTwoArrays349_3 {
    fun intersection(nums1: IntArray, nums2: IntArray): IntArray {
        nums1.sort()
        val result: MutableSet<Int> = hashSetOf()
        for (num in nums2) if (containsNum(nums1, num)) result.add(num)
        return result.toIntArray()
    }

    private fun containsNum(nums: IntArray, target: Int): Boolean {
        var left = 0
        var right = nums.size - 1
        while (left <= right) {
            val mid = left + (right - left) / 2
            if (nums[mid] == target) return true
            if (nums[mid] < target) left = mid + 1
            else right = mid - 1
        }
        return false
    }
}